regionServiceClient/lib/cloudregister/registerutils.py

```diff
@@ -0,0 +1,419 @@
+# License as published by the Free Software Foundation; either
+# version 3.0 of the License, or (at your option) any later version.
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.
+
+"""Utility functions for the cloud guest registration"""
+
+import base64
+import glob
+import logging
+import os
+import pickle
+import requests
+import stat
+import subprocess
+import sys
+
+from cloudregister import smt
+from M2Crypto import X509
+
+REGISTRATION_DATA_DIR = '/var/lib/cloudregister/'
+REGISTERED_SMT_SERVER_DATA_FILE_NAME = 'currentSMTInfo.obj'
+AVAILABLE_SMT_SERVER_DATA_FILE_NAME = 'availableSMTInfo_%d.obj'
+
+
+# ----------------------------------------------------------------------------
+def check_registration(smt_server_name):
+    """Check if the instance is already registerd"""
+    if has_repos(smt_server_name) and __has_credentials(smt_server_name):
+        return 1
+
+    return None
+
+
+# ----------------------------------------------------------------------------
+def clean_hosts_file(domain_name):
+    """Remove the smt server entry from the /etc/hosts file"""
+    new_hosts_content = []
+    content = open('/etc/hosts', 'r').readlines()
+    smt_announce_found = None
+    for entry in content:
+        if '# Added by SMT' in entry:
+            smt_announce_found = True
+            continue
+        if smt_announce_found and domain_name in entry:
+            smt_announce_found = False
+            continue
+        new_hosts_content.append(entry)
+
+    hosts_file = open('/etc/hosts', 'w')
+    for entry in new_hosts_content:
+        hosts_file.write(entry)
+    hosts_file.close()
+
+
+# ----------------------------------------------------------------------------
+def exec_subprocess(cmd):
+    """Execute the given command as a subprocess (blocking)"""
+    try:
+        cmd = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        cmd.communicate()
+    except:
+        return 0
+
+    return 1
+
+
+# ----------------------------------------------------------------------------
+def find_equivalent_smt_server(configured_smt, known_smt_servers):
+    """Find an SMT server that is equivalent to the currently configured
+       SMT server, only consider responsive servers"""
+    for smt in known_smt_servers:
+        if smt.get_ip() == configured_smt.get_ip():
+            continue
+        if smt.is_equivalent(configured_smt) and smt.is_responsive():
+            return smt
+
+    return None
+
+
+# ----------------------------------------------------------------------------
+def get_available_smt_servers():
```

Comment:
Looking now at how this is consumed by ` regionServiceClient/usr/lib/zypp/plugins/services/cloud_update`, wouldn't it make more sense to query the regionserver again, instead of relying on the cached document, in case any of the data has changed?